chain/network/src/concurrency/ctx/mod.rs

```diff
@@ -0,0 +1,153 @@
+    /// When Inner gets dropped, the context gets cancelled, so that
+    /// the tokio task which propagates the cancellation from
+    /// parent to child it terminated immediately and therefore doesn't
+    /// leak memory (see `Ctx::with_deadline`). However we don't want
+    /// the context to get cancelled only because the references to parent
+    /// are dropped. Therefore we keep a reference to parent here, so
+    /// that the parent is not dropped until all its children are dropped.
+    _parent: Option<Arc<Inner>>,
+}
+
+impl Drop for Inner {
+    fn drop(&mut self) {
+        // This will wake the task propagating cancellation to children (see `Ctx::with_deadline`).
+        // Note that since children keep a reference to the parent, at this point nobody awaits
+        // the cancelation of any child.
+        self.canceled.send();
+    }
+}
+
+/// See `Ctx::wait`.
+pub struct ErrCanceled;
+
+/// See `Ctx::wait`.
+pub type OrCanceled<T> = Result<T, ErrCanceled>;
+
+/// Ctx is an implementation of https://pkg.go.dev/context.
+/// Ctxs are a mechanism of broadcasting cancelation to concurrent routines (aka tokio tasks).
+/// The routines are expected to react to context cancelation on their own (they are not preempted)
+/// and perform a graceful shutdown. A tokio task is expected to return immediately once the context
+/// is cancelled (i.e. a task with canceled context, when polled in a loop should complete
+/// eventually without requiring any wake()).
+///
+/// Ctx is similar to a rust lifetime but for runtime:
+/// Ctx is expected to be passed down the call stack and to the spawned tokio subtasks.
+/// At any level of the call stack the Ctx can be narrowed down via `Ctx::with_cancel` (which
+/// allows the routine to cancel context of subroutines it spawns) or via
+/// `Ctx::with_timeout`/`Ctx::with_deadline`, which cancels the scope automatically after a given time.
+/// If is NOT possible to extend the context provided by the caller - a subtask is expected
+/// to adhere to the lifetime of its context and finish as soon as it gets canceled (or earlier).
+#[derive(Clone)]
+pub struct Ctx(Arc<Inner>);
+
+impl Ctx {
+    /// Constructs a new context:
+    /// * without a parent
+    /// * with real clock
+    /// * with infinite deadline
+    /// It should be called directly from main.rs.
+    pub fn inf() -> Ctx {
+        return Ctx(Arc::new(Inner {
+            canceled: signal::Once::new(),
+            deadline: time::Deadline::Inf,
+            clock: time::Clock::real(),
+            _parent: None,
+        }));
+    }
+
+    pub fn clock(&self) -> &time::Clock {
+        &self.0.clock
+    }
+
+    fn cancel(&self) {
+        self.0.canceled.send();
+    }
+
+    /// Check if the context has been canceled.
+    pub fn is_canceled(&self) -> bool {
+        self.0.canceled.try_recv()
+    }
+
+    /// Awaits for the context to get canceled.
+    /// Cancellable (in the rust sense).
+    pub async fn canceled(&self) {
```

Comment:
Ow, I thought that this is a 0-cost abstraction, since one basically compiles to another. Isn't that the case?

Reply:
Runtime wise it might be 0-cost in most cases due to inlining and such, however desugaring wise these aren’t equivalent, no.

---

```
async fn apple() {
    seeds.await;
}
```

is actually a

```
fn apple() -> impl Future<Output=()> {
    async {
        seeds.await;
    }
}
```

the slight difference is that every time poll happens, there will be a stack of two calls: first to the future created within `apple`, and then the `seeds` future itself. Returning the `seeds` directly removes this one layer of futures here.

---

Note that I emphasized the *may* here – this is a perfect example of a micro-optimization, and not even of the runtime (except maybe if we’re running debug builds?), but rather the compile time (less code for rustc to chew through).

Reply:
I'm not convinced that it makes any difference, but if you want it then sure.

Reply:
done